File: sim/orca.py
"""ORCA local collision avoidance (van den Berg, Guy, Lin and Manocha 2011), agents only.

The serial reference for the engine's navigation tiers (unity/.../Runtime/Orca.cs) and its
correctness oracle. Each neighbour contributes the half-plane of velocities that avoid it for
`tau` seconds if both agents take half the responsibility; the new velocity is the one closest to
the preferred velocity inside the speed circle and all half-planes (a 2D incremental linear
program), or, when those are infeasible, the one that violates them least (the RVO2
formulation: linear_program_3 over the projected constraints).

Neighbours are the k nearest within `reach`, nearest first, ties by index, so a port that finds
them with a grid picks exactly the same set. Everything is float64.
"""
from dataclasses import dataclass

import numpy as np

EPS = 1e-5
RADIUS = 0.3        # m, body radius
REACH = 2.5         # m, neighbour search radius (one engine grid cell)
TAU = 1.5           # s, time horizon


@dataclass
class Line:
    point: np.ndarray
    direction: np.ndarray


def _det(a, b):
    return a[0] * b[1] - a[1] * b[0]


def neighbours(pos, i, k, reach=REACH):
    """Indices of the k nearest agents to i within reach, nearest first, ties by index."""
    d = pos - pos[i]
    d2 = np.einsum("ij,ij->i", d, d)
    cand = np.flatnonzero((d2 < reach * reach) & (np.arange(len(pos)) != i))
    order = np.lexsort((cand, d2[cand]))
    return cand[order[:k]]


def half_planes(p, v, pn, vn, radius=RADIUS, tau=TAU, dt=1.0 / 30.0):
    """ORCA half-planes of an agent at p with velocity v against neighbours pn, vn."""
    lines = []
    inv_tau = 1.0 / tau
    r = 2.0 * radius
    r2 = r * r
    for pb, vb in zip(pn, vn):
        rel_pos = pb - p
        rel_vel = v - vb
        dist2 = rel_pos @ rel_pos
        if dist2 > r2:
            w = rel_vel - inv_tau * rel_pos
            w2 = w @ w
            dot1 = w @ rel_pos
            if dot1 < 0.0 and dot1 * dot1 > r2 * w2:
                # the velocity obstacle's cut-off circle
                wl = np.sqrt(w2)
                unit = w / wl
                direction = np.array([unit[1], -unit[0]])
                u = (r * inv_tau - wl) * unit
            else:
                # one of its legs
                leg = np.sqrt(dist2 - r2)
                if _det(rel_pos, w) > 0.0:
                    direction = np.array([rel_pos[0] * leg - rel_pos[1] * r, rel_pos[0] * r + rel_pos[1] * leg]) / dist2
                else:
                    direction = -np.array([rel_pos[0] * leg + rel_pos[1] * r, -rel_pos[0] * r + rel_pos[1] * leg]) / dist2
                u = (rel_vel @ direction) * direction - rel_vel
        else:
            # already overlapping: resolve within one step
            inv_dt = 1.0 / dt
            w = rel_vel - inv_dt * rel_pos
            wl = np.sqrt(w @ w)
            unit = w / wl
            direction = np.array([unit[1], -unit[0]])
            u = (r * inv_dt - wl) * unit
        lines.append(Line(v + 0.5 * u, direction))
    return lines


def _lp1(lines, n, radius, opt, direction_opt):
    ln = lines[n]
    dot = ln.point @ ln.direction
    disc = dot * dot + radius * radius - ln.point @ ln.point
    if disc < 0.0:
        return None
    s = np.sqrt(disc)
    t_left, t_right = -dot - s, -dot + s
    for i in range(n):
        den = _det(ln.direction, lines[i].direction)
        num = _det(lines[i].direction, ln.point - lines[i].point)
        if abs(den) <= EPS:
            if num < 0.0:
                return None
            continue
        t = num / den
        if den >= 0.0:
            t_right = min(t_right, t)
        else:
            t_left = max(t_left, t)
        if t_left > t_right:
            return None
    if direction_opt:
        return ln.point + (t_right if opt @ ln.direction > 0.0 else t_left) * ln.direction
    t = ln.direction @ (opt - ln.point)
    return ln.point + min(max(t, t_left), t_right) * ln.direction


def _lp2(lines, radius, opt, direction_opt):
    if direction_opt:
        result = opt * radius
    elif opt @ opt > radius * radius:
        result = opt / np.sqrt(opt @ opt) * radius
    else:
        result = opt.copy()
    for i in range(len(lines)):
        if _det(lines[i].direction, lines[i].point - result) > 0.0:
            r = _lp1(lines, i, radius, opt, direction_opt)
            if r is None:
                return i, result
            result = r
    return len(lines), result


def _lp3(lines, begin, radius, result):
    distance = 0.0
    for i in range(begin, len(lines)):
        if _det(lines[i].direction, lines[i].point - result) > distance:
            proj = []
            for j in range(i):
                det = _det(lines[i].direction, lines[j].direction)
                if abs(det) <= EPS:
                    if lines[i].direction @ lines[j].direction > 0.0:
                        continue
                    point = 0.5 * (lines[i].point + lines[j].point)
                else:
                    point = lines[i].point + (_det(lines[j].direction, lines[i].point - lines[j].point) / det) * lines[i].direction
                d = lines[j].direction - lines[i].direction
                proj.append(Line(point, d / np.sqrt(d @ d)))
            opt = np.array([-lines[i].direction[1], lines[i].direction[0]])
            fail, r = _lp2(proj, radius, opt, True)
            if fail >= len(proj):
                result = r
            distance = _det(lines[i].direction, lines[i].point - result)
    return result


def new_velocity(p, v, pref, max_speed, pn, vn, radius=RADIUS, tau=TAU, dt=1.0 / 30.0):
    lines = half_planes(p, v, pn, vn, radius, tau, dt)
    fail, result = _lp2(lines, max_speed, pref, False)
    if fail < len(lines):
        result = _lp3(lines, fail, max_speed, result)
    return result


def step(pos, vel, pref, max_speed, k, dt=1.0 / 30.0, reach=REACH):
    """One synchronous (Jacobi) ORCA step for every agent: new velocities from the snapshot."""
    out = np.empty_like(vel)
    for i in range(len(pos)):
        nb = neighbours(pos, i, k, reach)
        out[i] = new_velocity(pos[i], vel[i], pref[i], max_speed[i], pos[nb], vel[nb], dt=dt)
    return out

File: unity/Parity3D/Assets/Scripts/Runtime/Orca.cs
// ORCA local collision avoidance (van den Berg, Guy, Lin and Manocha 2011), agents only: the
// port of sim/orca.py, which is its oracle (tools/gen_oracle_cases.py, orca section). Doubles
// throughout so the port and the oracle agree to rounding. One instance per thread: the line
// buffers are its only state.
using System;

namespace Parity
{
    public sealed class Orca
    {
        public const double Eps = 1e-5;
        public const double Radius = 0.3;   // m, body radius
        public const double Reach = 2.5;    // m, neighbour search radius (one grid cell)
        public const double Tau = 1.5;      // s, time horizon

        struct Line { public double Px, Py, Dx, Dy; }

        Line[] lines, proj;

        public Orca(int maxNeighbours)
        {
            lines = new Line[Math.Max(maxNeighbours, 1)];
            proj = new Line[Math.Max(maxNeighbours, 1)];
        }

        static double Det(double ax, double ay, double bx, double by) => ax * by - ay * bx;

        /// <summary>The new velocity of an agent at p moving at v, preferring pref, against m
        /// neighbours (positions nx, ny, velocities nvx, nvy).</summary>
        public void NewVelocity(double px, double py, double vx, double vy, double prefX, double prefY, double maxSpeed,
                                double[] nx, double[] ny, double[] nvx, double[] nvy, int m, double dt,
                                out double rx, out double ry)
        {
            if (lines.Length < m) { lines = new Line[m]; proj = new Line[m]; }
            double invTau = 1.0 / Tau, r = 2.0 * Radius, r2 = r * r;
            for (int j = 0; j < m; j++)
            {
                double relX = nx[j] - px, relY = ny[j] - py;
                double rvX = vx - nvx[j], rvY = vy - nvy[j];
                double dist2 = relX * relX + relY * relY;
                double dirX, dirY, uX, uY;
                if (dist2 > r2)
                {
                    double wX = rvX - invTau * relX, wY = rvY - invTau * relY;
                    double w2 = wX * wX + wY * wY;
                    double dot1 = wX * relX + wY * relY;
                    if (dot1 < 0.0 && dot1 * dot1 > r2 * w2)
                    {
                        // the velocity obstacle's cut-off circle
                        double wl = Math.Sqrt(w2), ux = wX / wl, uy = wY / wl;
                        dirX = uy; dirY = -ux;
                        uX = (r * invTau - wl) * ux; uY = (r * invTau - wl) * uy;
                    }
                    else
                    {
                        // one of its legs
                        double leg = Math.Sqrt(dist2 - r2);
                        if (Det(relX, relY, wX, wY) > 0.0)
                        {
                            dirX = (relX * leg - relY * r) / dist2; dirY = (relX * r + relY * leg) / dist2;
                        }
                        else
                        {
                            dirX = -(relX * leg + relY * r) / dist2; dirY = -(-relX * r + relY * leg) / dist2;
                        }
                        double dot2 = rvX * dirX + rvY * dirY;
                        uX = dot2 * dirX - rvX; uY = dot2 * dirY - rvY;
                    }
                }
                else
                {
                    // already overlapping: resolve within one step
                    double invDt = 1.0 / dt;
                    double wX = rvX - invDt * relX, wY = rvY - invDt * relY;
                    double wl = Math.Sqrt(wX * wX + wY * wY), ux = wX / wl, uy = wY / wl;
                    dirX = uy; dirY = -ux;
                    uX = (r * invDt - wl) * ux; uY = (r * invDt - wl) * uy;
                }
                lines[j] = new Line { Px = vx + 0.5 * uX, Py = vy + 0.5 * uY, Dx = dirX, Dy = dirY };
            }
            int fail = Lp2(lines, m, maxSpeed, prefX, prefY, false, out rx, out ry);
            if (fail < m) Lp3(m, fail, maxSpeed, ref rx, ref ry);
        }

        static bool Lp1(Line[] ls, int n, double radius, double optX, double optY, bool dirOpt, out double rx, out double ry)
        {
            rx = ry = 0.0;
            var ln = ls[n];
            double dot = ln.Px * ln.Dx + ln.Py * ln.Dy;
            double disc = dot * dot + radius * radius - (ln.Px * ln.Px + ln.Py * ln.Py);
            if (disc < 0.0) return false;
            double s = Math.Sqrt(disc), tLeft = -dot - s, tRight = -dot + s;
            for (int i = 0; i < n; i++)
            {
                double den = Det(ln.Dx, ln.Dy, ls[i].Dx, ls[i].Dy);
                double num = Det(ls[i].Dx, ls[i].Dy, ln.Px - ls[i].Px, ln.Py - ls[i].Py);
                if (Math.Abs(den) <= Eps)
                {
                    if (num < 0.0) return false;
                    continue;
                }
                double t = num / den;
                if (den >= 0.0) tRight = Math.Min(tRight, t); else tLeft = Math.Max(tLeft, t);
                if (tLeft > tRight) return false;
            }
            double tt;
            if (dirOpt) tt = optX * ln.Dx + optY * ln.Dy > 0.0 ? tRight : tLeft;
            else tt = Math.Min(Math.Max(ln.Dx * (optX - ln.Px) + ln.Dy * (optY - ln.Py), tLeft), tRight);
            rx = ln.Px + tt * ln.Dx; ry = ln.Py + tt * ln.Dy;
            return true;
        }

        static int Lp2(Line[] ls, int m, double radius, double optX, double optY, bool dirOpt, out double rx, out double ry)
        {
            if (dirOpt) { rx = optX * radius; ry = optY * radius; }
            else if (optX * optX + optY * optY > radius * radius)
            {
                double l = Math.Sqrt(optX * optX + optY * optY);
                rx = optX / l * radius; ry = optY / l * radius;
            }
            else { rx = optX; ry = optY; }
            for (int i = 0; i < m; i++)
                if (Det(ls[i].Dx, ls[i].Dy, ls[i].Px - rx, ls[i].Py - ry) > 0.0)
                {
                    if (!Lp1(ls, i, radius, optX, optY, dirOpt, out double nx, out double ny)) return i;
                    rx = nx; ry = ny;
                }
            return m;
        }

        void Lp3(int m, int begin, double radius, ref double rx, ref double ry)
        {
            double distance = 0.0;
            for (int i = begin; i < m; i++)
            {
                var li = lines[i];
                if (Det(li.Dx, li.Dy, li.Px - rx, li.Py - ry) <= distance) continue;
                int np = 0;
                for (int j = 0; j < i; j++)
                {
                    var lj = lines[j];
                    double det = Det(li.Dx, li.Dy, lj.Dx, lj.Dy);
                    double ptX, ptY;
                    if (Math.Abs(det) <= Eps)
                    {
                        if (li.Dx * lj.Dx + li.Dy * lj.Dy > 0.0) continue;
                        ptX = 0.5 * (li.Px + lj.Px); ptY = 0.5 * (li.Py + lj.Py);
                    }
                    else
                    {
                        double t = Det(lj.Dx, lj.Dy, li.Px - lj.Px, li.Py - lj.Py) / det;
                        ptX = li.Px + t * li.Dx; ptY = li.Py + t * li.Dy;
                    }
                    double dx = lj.Dx - li.Dx, dy = lj.Dy - li.Dy, dl = Math.Sqrt(dx * dx + dy * dy);
                    proj[np++] = new Line { Px = ptX, Py = ptY, Dx = dx / dl, Dy = dy / dl };
                }
                if (Lp2(proj, np, radius, -li.Dy, li.Dx, true, out double nx, out double ny) >= np) { rx = nx; ry = ny; }
                distance = Det(li.Dx, li.Dy, li.Px - rx, li.Py - ry);
            }
        }

        /// <summary>Keeps the k nearest of a stream of candidates, nearest first, ties by index --
        /// the order sim/orca.py::neighbours returns. Returns the new count.</summary>
        public static int Insert(int[] idx, double[] d2, int count, int k, int j, double dj)
        {
            int at = count;
            while (at > 0 && (d2[at - 1] > dj || (d2[at - 1] == dj && idx[at - 1] > j))) at--;
            if (at >= k) return count;
            int last = Math.Min(count, k - 1);
            for (int s = last; s > at; s--) { idx[s] = idx[s - 1]; d2[s] = d2[s - 1]; }
            idx[at] = j; d2[at] = dj;
            return Math.Min(count + 1, k);
        }
    }
}
